x/gov/handler.go

```diff
@@ -0,0 +1,105 @@
+package gov
+
+import (
+	"time"
+
+	"github.com/iov-one/weave"
+	"github.com/iov-one/weave/errors"
+	"github.com/iov-one/weave/x"
+)
+
+const (
+	newTextProposalCost = 0
+	voteCost            = 0
+	tallyCost           = 0
+)
+
+// RegisterQuery registers governance buckets for querying.
+func RegisterQuery(qr weave.QueryRouter) {
+	NewElectionRulesBucket().Register("electionrules", qr)
+	NewElectorateBucket().Register("electorates", qr)
+	NewProposalBucket().Register("proposal", qr)
+}
+
+// RegisterRoutes registers handlers for governance message processing.
+func RegisterRoutes(r weave.Registry, auth x.Authenticator) {
+	propBucket := NewProposalBucket()
+	elecBucket := NewElectorateBucket()
+	r.Handle(pathVoteMsg, &VoteHandler{
+		auth:       auth,
+		propBucket: propBucket,
+		elecBucket: elecBucket,
+	})
+}
+
+type VoteHandler struct {
+	auth       x.Authenticator
+	elecBucket *ElectorateBucket
+	propBucket *ProposalBucket
+}
+
+func (h VoteHandler) Check(ctx weave.Context, db weave.KVStore, tx weave.Tx) (weave.CheckResult, error) {
+	var res weave.CheckResult
+	if _, _, _, err := h.validate(ctx, db, tx); err != nil {
+		return res, err
+	}
+	res.GasAllocated += voteCost
+	return res, nil
+
+}
+
+func (h VoteHandler) Deliver(ctx weave.Context, db weave.KVStore, tx weave.Tx) (weave.DeliverResult, error) {
+	var res weave.DeliverResult
+	vote, proposal, elector, err := h.validate(ctx, db, tx)
+	if err != nil {
+		return res, err
+	}
+	if err := proposal.Vote(vote.Selected, *elector); err != nil {
+		return res, err
+	}
+	// todo: set tag to mark voting ???
```

Comment:
Does not hurt!

Reply:
In cosmos-sdk only proposal events are tagged but not the votes. I would go with that for now as votes on different proposals in the same block may require more work due the overwrite case.